site/content/entry/citra-progress-report-2017-q4.md

```diff
@@ -0,0 +1,202 @@
+
+A little known feature in macOS is fat binaries, the ability to have the same executable
+contain binaries for multiple architectures. This was used extensively around 2007
+to support Mac OS X's transition from PowerPC to x86_64, allowing developers to
+have one binary work on all Macs effortlessly.
+
+After poking around Apple's LLVM fork, [MerryMage](https://github.com/MerryMage)
+found that they had a specific flag to build binaries for Intel Haswell and later.
+Enabling it made a fat binary that contained both pre-Haswell and post-Haswell code,
+allowing Citra to take advantage of the newer CPU's instructions, without dropping
+support for older ones.
+
+## [Kernel/IPC: Add a small delay after each SyncRequest to prevent thread starvation.](https://github.com/citra-emu/citra/pull/3091) by [Subv](https://github.com/Subv)
+
+When communicating with services, titles on a real Nintendo 3DS expect to wait some
+amount of time before the service replies or responds. Until now, Citra didn't implement
+this, it responded without incrementing the virtual clock. From the emulated system's
+perspective, it appeared as though services replied and reponded literally instantly
+because of this, leading to strange side-effects.
+
+For example, some buildings in *Animal Crossing: New Leaf* fail to render and
+cannot be interacted with. Or worse, some games such as *Star Fox 3D* can't even
+reach the title screen because of this.
+
+{{< figure src="/images/entry/citra-progress-report-2017-q4/fsUSER-threads.png" 
+    title="A demonstration of how instant service replies can break a title" >}}
+
+The example in this figure is taken directly from *Star Fox 3D*. When the game boots,
+it checks if any save data exists on the console's NAND, and tries to recreate it
+if it doesn't. It does this by sending a request to the service responsible for
+file access, and then waiting for a reply. Because the request is synchronous,
+the thread that made the request is put to sleep and its priority lowered. The
+game actually uses this time that would otherwise be spent waiting by having other
+threads run while the original thread was still asleep, so that by the time it
+woke up with the response, it would have other resources it needed ready to go.
+
+Because Citra's responses are instantaneous from the point-of-view of the game,
+the secondary thread doesn't have nearly enough time to finish its job. Because
+the first thread's priority comes back up above the secondary's once it wakes up,
+this leads to the first thread waiting on the secondary thread to finish, but it
+never gets a chance to, because of its now much lower priority. Essentially, the
+first thread waits on the secondary forever, because the secondary never gets a
+chance to actually finish what it was doing.
+
+After [B3n30](https://github.com/B3n30) did some testing with homebrew written by
```

Comment:
ds wrote the homebrew to show there was a problem. subv wrote a homebrew to get some values to test which b3n30 ran